Improve decidim's i18n

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

This PR fixes several things regarding decidim languages:

* ~Adds docs! :tada:~

* Allows configuring locales. At the moment, by the time decidim's configuration is run, the `config.available_locales = Decidim.available_locales` assignment in `application.rb` has already been run. So custom locales are not picked up. I fixed it by changing the order these are run.

* ~Changes generated application locales to be `:en`, `:ca` & `:es`, instead of all available locales. At the moment, playing around with the generated application is a PITA because until we fix #1127, all languages are required. And the generated application uses all 8 languages by default! Even if #1127 was no longer a problem, I'm not sure we want to generate applications with all languages enabled, since it would be a pretty unrealistic application. I think enabling English + decidim original languages is a good compromise.~. I finally used those locales not globally but for the sample organizations as suggested by @mrcasals.

* ~Makes it easier to remove locales from an app by providing a rake task, `decidim:sanitize_locales`, that cleans up DB from no longer available locales. Without running this task, decidim can crash after removing locales, because the locales available for each organization might no longer be available.~

#### :pushpin: Related Issues
- ~Fixes #1439~.
- Related to #1495. ~Doesn't completely fix it, but gets it started.~
- ~Blocked on #1554, since decidim can't provide tasks to applications until that one is merged. It's currently built on top of it.~ That PR was merged.

#### :clipboard: Subtasks
_None_.

### :camera: Screenshots (optional)
_None_.

#### :ghost: GIF
![catcold](https://user-images.githubusercontent.com/2887858/27731508-761fef4c-5d8d-11e7-936f-67cec6b86904.gif)

mrcasals commented on lib/generators/decidim/templates/initializer.rb:
```diff
@@ -7,7 +7,7 @@
   config.authorization_handlers = [ExampleAuthorizationHandler]
 
   # Uncomment this lines to set your preferred locales
-  # config.available_locales = %i{en ca es}
```
Since Ruby 2.0 this is equivalent

deivid-rodriguez replied:
It is. But we are enforcing the style I switched to via [rubocop](https://github.com/decidim/decidim/blob/fa4755756c85041671bf992a3541d7f1ebb74525/.rubocop.yml#L939-L944).

mrcasals commented:
@deivid-rodriguez the second point can be related to #1532, if you want to link it :)

mrcasals commented on lib/generators/decidim/templates/initializer.rb:
```diff
@@ -6,8 +6,8 @@
-  # Uncomment this lines to set your preferred locales
-  # config.available_locales = [:en, :ca, :es]
+  # Change this line to set your preferred locales
+  config.available_locales = [:en, :ca, :es]
```
Maybe the problems are the seeds, then? Since the organization created there uses `Decidim.available_locales` as its own `available_locales`. Maybe we can change it there?

deivid-rodriguez replied:
You mean using the hardcoded list for the organizations locales, but leaving the full list enabled here? As long as I don't have to fill in 8 locales for each field on every form, sounds good to me! :)

mrcasals replied:
Yep, I'd change the values in the seeds until we start working on not making all the languages compulsory

mrcasals commented on decidim-dev/lib/tasks/locale_checker.rake:
```diff
@@ -18,4 +18,14 @@ namespace :decidim do
+  task sanitize_locales: :environment do
+    Decidim::Organization.find_each do |organization|
+      sanitized_locales =
+        organization.available_locales & Decidim.available_locales.map(&:to_s)
```
I'm not sure about this. Shouldn't we update processes, etc in order to fix the locales there too?

deivid-rodriguez replied:
I guess, I didn't know there were other locales in DB.

deivid-rodriguez replied:
If we configure a reasonable set of locales per organization, I don't think we need this at the moment. I'll remove it, ok?

mrcasals replied:
👌

deivid-rodriguez replied:
Now I see what you meant by updating processes. All translations for every text field are saved in DB, of course...

mrcasals replied:
Yeah, that's what I meant. We would need to update every process/meeting/whatever and remove that language :/

deivid-rodriguez replied:
Well, I'm not sure leaving the translations around could cause any problems. And they would still be there if you reenable the language... :) I can see how the opposite is problematic in the current situation though (adding a new language and every object lacking _compulsory_ translations).

mrcasals commented on docs/getting_started.md:
```diff
@@ -75,6 +75,7 @@ We also have other guides on how to configure some extra features:
 - [Social providers integration](https://github.com/decidim/decidim/blob/master/docs/social_providers.md): Enable sign up from social networks.
 - [Analytics](https://github.com/decidim/decidim/blob/master/docs/analytics.md): How to enable analytics
 - [Geocoding](https://github.com/decidim/decidim/blob/master/docs/geocoding.md): How to enable geocoding for proposals and meetings
+- [Languages](https://github.com/decidim/decidim/blob/master/docs/languages.md): How to enable configure languages
```
`How to configure languages`?

deivid-rodriguez commented:
@mrcasals Your comments make sense to me, it should make this much simpler. I'll update the PR & docs as you suggested.

deivid-rodriguez commented:
> Allows configuring locales. At the moment, by the time decidim's configuration is run, the config.available_locales = Decidim.available_locales assignment in application.rb has already been run. So custom locales are not picked up. I fixed it by changing the order these are run.

Should I keep 32c00fb? Removing locales globally is currently prone to errors (you have to re-seed or use a task like the one I provided here), and provides little value over configuring them per organization. So I'm not even sure about keeping that either, even if it fixes unexpected behavior.

mrcasals commented:
We deliberately don't let system admins modify available locales for an already created organization, because we thought that modifying the locales could give some problems right now (until we do that PR about the locales). 

As for that commit, I'd ping @beagleknight so he can chime in :)

deivid-rodriguez commented:
Oh, I didn't know that... #1127 keeps biting! :scream:

deivid-rodriguez commented:
I rebased this PR and removed the parts I was not sure about. Updated the PR description to reflect what's left in there. It's ready for another review pass!

deivid-rodriguez commented:
Anything else left to do here? This makes manual testing a lot easier IMO.

deivid-rodriguez commented:
Yuhuuuuu! :tada:

beagleknight commented on decidim-core/db/seeds.rb:
```diff
@@ -17,7 +17,7 @@
     homepage_image: File.new(File.join(__dir__, "seeds", "homepage_image.jpg")),
     default_locale: I18n.default_locale,
-    available_locales: Decidim.available_locales,
+    available_locales: [:en, :ca, :es],
```
Hi @deivid-rodriguez ! I know this is merged but I am having a few problems because of that 😄 . Do you remember why you hardcoded this values? 
The problem I found is when a installation doesn't have English for example, I cannot create a review app because the seeds are creating "invalid" organizations.

beagleknight replied:
No problem! 😄  I think I am going to change our app generator to include those locales as default because if we leave `Decidim.available_locales` commented it will use all the locales.

deivid-rodriguez replied:
Sorry about that... :pray: I can rework another PR that doesn't touch the seed process if you want.

deivid-rodriguez replied:
Heheh, we replied at the same time :) Sounds good to me, sorry for the trouble :sweat_smile:

beagleknight replied:
Don't worry! I think I can handle it and it's not a big deal 😄
